### PR TITLE
Implement optimization described in #92

### LINKED ISSUE #92: lyse speedup and the analysis plot panel
Hey everyone,

For our experiment at EPFL, we were writing a plotting library [analysis_plot_panel](https://lqg-epfl.github.io/analysis_plot_panel/)  that relies on the pyqtgraph. Our experimental rate is higher than 1 Hz. This means that we had to streamline our analysis scripts. The issue was that Matplolib takes a significant time (sometimes more than 1s) to update simple plots. This meant that we had to deactivate the plots in some cases. This is of course not optimal when one whats to debug an experiment.

The analysis_plot_panel solves this issue since pyqtgraph allows for high-speed refreshes (>100 Hz) for 1d and 2d figures. The script is integrated into lyse using a single shot routine. Check the [documentation](https://lqg-epfl.github.io/analysis_plot_panel/) for details.

The reason we are sharing this here as an issue is, that during our research to find out how to speedup lyse, we learned that a lot of the time of the analysis is wasted at the loading and saving stage (from h5_file). The reason for this was not the data transfer, but the fact, that at every save_result call the h5 file would be opened and closed again. This takes significantly more time than saving the actual result. We solved this issue by rewriting the functions in the __init__.py file. Now they take h5_file as an argument. This means that if you call save_results it opens the file only once and hands the file to all of the save_result calls. Would there be a possibility to integrate this feature into lyse? The new __init__.py file can be found [here](https://github.com/LQG-EPFL/analysis_plot_panel/blob/master/new_lyse_init/__init__.py).

### PR DESCRIPTION
**Pull request by dihm**
Implements a wrapper function that allows all lyse functions to optionally take a handle to the h5_file instead of opening and closing for each operation. This allows a single open and close operation for a single shot run.

This tends to speed up writes more than reads.

This closes #92

### CODE REVIEW

**Comment by dihm** [comment]
Using a very simple test analysis script with crude timing seems to suggest that reads get little to modest improvements (up to x1.2, potentially being elided by numpy magic), and writes get a more significant improvement (x20-x40).

```python
from lyse import *
import time

run = Run(path)

start = time.time()
for _ in range(10):
    _, _ = run.get_trace('ClockTest')

end = time.time()
orig_read = end-start
print(f'Elapsed original read: {orig_read:.3e} s')

start = time.time()
with h5py.File(run.h5_path, 'r') as f:
    
    for _ in range(10):
        _, _ = run.get_trace('ClockTest', h5_file=f)

end = time.time()

new_read = end - start
print(f'Elapsed new read: {new_read:.3e} s')
print(f'New is {orig_read/new_read:.3f} times faster')


start = time.time()
for _ in range(100):
    run.save_result('ClockTest',5)

end = time.time()
orig_write = end-start
print(f'Elapsed original write: {orig_write:.3e} s')

start = time.time()
with h5py.File(run.h5_path, 'r+') as f:

    for _ in range(100):
       run.save_result('ClockTest', 5, h5_file=f)

end = time.time()
new_write = end-start
print(f'Elapsed new write: {new_write:.3e} s')
print(f'New is {orig_write/new_write:.3f} times faster')
```

A typical output from this script is
```shell
Elapsed original read: 8.700e-01 s
Elapsed new read: 8.535e-01 s
New is 1.019 times faster
Elapsed original write: 7.203e-01 s
Elapsed new write: 1.997e-02 s
New is 36.076 times faster
```

**Comment by dihm** [comment]
Assuming there aren't objections, I'd like to add something to the documentation describing how to do this more clearly before merging.

**Comment by philipstarkey** [comment]
I think a better way to do this might be to add method to the run class that is decorated with `contextmanager` that opened the file. Then you could store the open file on the run instance, detect it's open, and not have to pass it in. Should make adapting existing analysis code easier as method signature would change, you just need to add a `with run.open('w'):` or something like that to your analysis code.

**Comment by dihm** [comment]
That makes sense. After editing three method signatures I had a quick moment of "there must be a better way" then promptly moved on just so I could have something to test that the speed-ups are actually worth it. I'll give it a better crack this time and hopefully the result will be a bit more user friendly.

**Comment by dihm** [comment]
I've had another pass at it. One nice benefit to this method is that existing methods which make successive calls to the lyse API (ie `save_result_arrays`) now automatically benefit from this without any change on the user's part.

A potential downside is that mixing methods using the contextmanager with the older per-method implementation (eg local user modifications to extend lyse functionality with new Run.methods) is all but certain to lead to multiple attempts to open the same file. Users will be forced to update any local changes to lyse to match this style if they want to take advantage of the new feature.

**Comment by dihm** [comment]
@philipstarkey If you have another moment, I'd appreciate your eyes on this updated PR before I merge it.

Maybe my one question is whether the `open_file` wrapper is implemented correctly, particularly if it should actually be a method of the `Run` class?

**Comment by philipstarkey** [comment]
I'll try and have a look this weekend!

File: lyse/__init__.py
Comment (philipstarkey) [inline lyse/__init__.py:355]: This should probably also use a context manager. Something like:

```
with h5py.File(self.h5_path, mode) as f:
    self.__h5_file = f
    try:
        yield f
    finally:
        self.__h5_file = None
```

That way you both yield the file object like h5py.File does, and you leave it up the the h5py context manager to close the file properly.

File: lyse/__init__.py
Comment (philipstarkey) [inline lyse/__init__.py:343]: This could be condensed to `if self.no_write and mode != “r”:`

File: lyse/__init__.py
Comment (philipstarkey) [inline lyse/__init__.py:348]: If we new the mode wasn’t changing, there is nothing stopping us yielding `self.__h5_file` here (and not closing the file afterwards). Does that go some way to helping with resuable code snippets that might not know if they’re being called inside the context manager already?

File: lyse/__init__.py
Comment (philipstarkey) [inline lyse/__init__.py:357]: I think this decorator doesn’t need to be inside the `Run` class? Probably clearer if it’s not (otherwise it’s odd that it’s not defined as a `staticmethod` and also doesn’t take `self` as an argument)

File: lyse/__init__.py
Comment (philipstarkey) [inline lyse/__init__.py:382]: If statement can also be condensed as above

File: lyse/__init__.py
Comment (philipstarkey) [inline lyse/__init__.py:392]: I think some additional documentation around this (either comment here or in docstring) would be useful. It took me a while to work out that passing in `r+` as the mode (e.g. see use of decorator further down file) wasn’t incompatible with someone manually opening the file with `a` for example. I think this is something that will be easily forgotten unless we document it now. The only incompatible case is the one for this error message, where if the file is open in read only, then any decorator use that wants write access needs to throw an error.

That could be an argument for not actually having the argument to the decorator be `mode`, but a `Boolean` that indicates whether we need it to be open in a writable mode or not. That boolean would then toggle between `r` or `a` in the call to `self.open()` a few lines above.

Comment (philipstarkey) [reply lyse/__init__.py:392]: Which brings me to another point, it seems like we’ve switched the use of `a` to `r+` which would be a backwards incompatible change?

File: lyse/__init__.py
Comment (philipstarkey) [inline lyse/__init__.py:666]: Unrelated to this PR, but can you change this to `group not in self.h5_file` while you are here? It is equivalent, but more readable syntax

Comment (philipstarkey) [reply lyse/__init__.py:666]: Looks like there are a couple of others too in some later methods

**Review by philipstarkey — CHANGES_REQUESTED** [review]
Looking good. Got some suggested changes. Hopefully not too nitpicky…I realised afterwards that I reviewed this like I would something at work 😅

File: lyse/__init__.py
Comment (dihm) [reply lyse/__init__.py:355]: It's context managers all the way down!

File: lyse/__init__.py
Comment (dihm) [reply lyse/__init__.py]: Well noted! This is the result of simplifying the even more convoluted logic tree I started with.

File: lyse/__init__.py
Comment (dihm) [reply lyse/__init__.py:354]: I thought that would be the case, just wanted to get another opinion.

Now that I look at it, I'm pleasantly surprised it even works as is. It's not like I called it as a typical class method. TIL more about what using self actually does.

File: lyse/__init__.py
Comment (dihm) [reply lyse/__init__.py:345]: Fair enough. Just need to make sure the modes are compatible, then return.

Comment (dihm) [reply lyse/__init__.py:345]: Actually, I don't think we have to yield the file handle (normal operation doesn't). If fact, I think it has to `return None` here to match signatures and prevent executing the file open, right?

File: lyse/__init__.py
Comment (dihm) [reply lyse/__init__.py:393]: Yeah, this is due to the quirks of how h5py File objects work. While there are a number of mode options for [opening the file](https://docs.h5py.org/en/stable/high/file.html#opening-creating-files), the file itself can only be in [two modes](https://docs.h5py.org/en/stable/high/file.html#h5py.File.mode) (`'r'` and `'r+'`). All the other options for opening the file deal exclusively with how to handle creating a new file. I'll update the docstring to make this clear as it isn't super intuitive at first glance.

This relates to my thought process for changing most of the calls from `a` to `r+`. The only difference between these two is that `a` will create the file if it doesn't exist already. It is my understanding that almost all lyse functions should never create a new file on their own: that job is handled by runmanager or the `Sequence` constructor. For example, if `save_result` is creating a new file, I'd argue that is a bug that should raise an error to be fixed instead of failing silently as it would now. Assuming that reasoning is valid, I see there are a few docstrings that should be updated to reflect that. I've gone ahead and changed them. If my reasoning is wrong, I can change things back.

I decided to keep mode strings (instead of a bool) to match the h5py documentation. Felt like it would be more clear what was happening if we didn't add another layer of abstraction over top.

File: lyse/__init__.py
Comment (dihm) [reply lyse/__init__.py]: I think I have found them all. There were quite a few.

File: lyse/__init__.py
Comment (dihm) [reply lyse/__init__.py]: Ah, should have tested this out before pushing. Naive putting the decorator outside the class [doesn't work](https://stackoverflow.com/questions/1263451/python-decorators-in-classes) because you need to access the class instance from within the wrapper and that doesn't work unless we define within the class. Not sure how much I trust a single SO post for best practices, but what I had appears to be a fairly common pattern. I'll revert for now.

**Comment by dihm** [comment]
Don't worry about nitpicking. I'm a serious nitpicker on reviews myself 😉 I appreciate the effort it takes to put extra eyes on someone else's code.

File: lyse/__init__.py
Comment (philipstarkey) [reply lyse/__init__.py:356]: This really should have worked. `self` isn't accessed until the method is called (e.g. it's used inside `wraps` which is the replacement method returned by the decorator). It's equivalent to the `define_state` decorator in BLACS. The only difference is the use of the `functools.wraps`. that should be fine butmaybe I'm missing something. Is the use of wraps meant to be on the middle function instead? It's primary purpose is to copy the docstring of the `Run` methods decorated with `open`. Should be easy enough to reason out where it needs to be and test it's copying the doc strings

File: lyse/__init__.py
Comment (philipstarkey) [reply lyse/__init__.py:393]: A reason for `a` might be if you're using `Run` as a multishot analysis storage system. I suppose that's probably a niche use case and we can just require you create the file first in that case.

Thoughts @chrisjbillington ?

File: lyse/__init__.py
Comment (philipstarkey) [reply lyse/__init__.py]: You need to prevent the later open and yield occurring. I think you still want to use yield too though. Maybe:

```
    yield self.__h5_file
    return
```

Otherwise you don't get the file handle out the second time, so you get incompatible behaviour of you reuse some code inside an already open context manager. And I don't know how the `contextmanager` decorator handles no yield.

File: lyse/__init__.py
Comment (chrisjbillington) [reply lyse/__init__.py:393]: It is niche, and we shouldn't worry much about niche things.

Though for what it's worth, I also think worrying about someone passing in a non-existent file unintentionally is fairly niche. So I don't think it really matters.

On balance though, I lean toward discouraging using `Run` for multi-shot storage, because that means we have to continue to commit to ensuring it works with arbitrary HDF5 files, whereas it would be nice to be able to assume the HDF5 file is an actual shot file. It currently works with arbitrary files, but minimising future commitments is good.

And if someone wants to use HDF5 to store data, they may do so with `h5py` directly - our wrapper is not adding much and is actually pretty limiting. I would encourage using raw `h5py` instead for this, so I have no problem making `Run` less convenient to use for this, it is unlikely to be the best tool for the job anyway.

So +1 for `r+` I think.

File: lyse/__init__.py
Comment (dihm) [reply lyse/__init__.py]: Well I can report that contextManager doesn't like it if there isn't a yield since it is a generator. TIL.

I'm still confused why it should yield the file handle instead of `None`. The file handle isn't returned by this context normally; it just saves it as a Run attribute.

I suppose that means we can't do `with run.open('r') as f`, but my working assumption is that we didn't want people to normally access the file directly. If they really need the handle, they can get it from the `run.h5_file` property. I think I slightly prefer that (as it discourages casual direct interaction with the h5 file since you need to separately get the handle) but I don't have strong feelings about it. Either is fine.

File: lyse/__init__.py
Comment (dihm) [reply lyse/__init__.py:394]: Cool. I'll keep as-is then.

Also, I'm pretty sure `Run` will currently fail if called on a non-existing file thanks to `_create_group_if_not_exists`, which already uses `r+`, so hopefully this won't break anyone's workflow.

https://github.com/labscript-suite/lyse/blob/82db03237f8667ba65859c1fb9e647a123bfb98e/lyse/__init__.py#L331-L341

File: lyse/__init__.py
Comment (philipstarkey) [reply lyse/__init__.py]: The reason to yield the file handle is so it mimics `h5py.File(..)`. Probably most people won't use it. And yep, you could access it via the attribute on the `Run` instance. But given we have to yield, we may as well yield something useful! You don't have to even assign it (`with run.open(mode)` should work as well as `with run.open(mode) as f`)

Comment (philipstarkey) [reply lyse/__init__.py]: Think I must have not read the whole comment before I replied. Feel like I've just repeated what you said. Oops. I don't see any harm in given people the file handle though. Sometimes you might want direct access to the file. Generally you expect context managers that open a file to yield the file handle as well. I think my feelings are mainly driven by that - trying to be consistent with how you'd expect it to behave.

File: lyse/__init__.py
Comment (philipstarkey) [reply lyse/__init__.py:357]: Oh, I see. Given you don't need to write to that attribute inside the decorator, can you just switch it to using the `self.h5_file` property instead? Then the decorator could live outside of the class.

Comment (dihm) [reply lyse/__init__.py]: OK, latest push should resolve this one too.

**Comment by dihm** [comment]
I think everything is addressed. Ready for a re-review at your leisure.

File: lyse/__init__.py
Comment (chrisjbillington) [reply lyse/__init__.py]: For what it's worth, the name mangling is achieving nothing and could be removed. The purpose of double-underscore mangling is to make attributes inaccessible from subclasses or parent classes, which we are not trying to do.

There's not even much reason to make them single-underscore private, Id' argue since they all have getters (I guess some don't have setters). My guess is that the main reason they are private at all is to make it easier to document them by adding docstrings to the getters.

But yeah, single-underscore is arguable, whereas double is achieving nothing.

Comment (philipstarkey) [reply lyse/__init__.py]: The name mangling, and exposure via property getters, is to prevent end users from changing the internal state. A lot of the checks, etc. rely on the assumption that these variables are private. If one of those mangled variables was changed (accidentally) I think you'd see very odd behaviour.

Comment (chrisjbillington) [reply lyse/__init__.py]: Well that's an argument for them being private generally (i.e. one leading underscore), which I lean against more than you but accept there's a case for. Name mangling as triggered by two leading underscores is for privacy from methods of subclasses and parent classes though, which I don't think there is a case for.

**Comment by philipstarkey** [comment]
I'm happy with this.

One thought I did have was whether it made more sense to `yield self` rather than the file handle. It's more useful, because you can then do `with Run(path).open("r+") as shot:`...not sure if it's more confusing though (not that you'd be forced to use it)

**Comment by dihm** [comment]
> One thought I did have was whether it made more sense to yield self rather than the file handle. It's more useful, because you can then do with Run(path).open("r+") as shot:...not sure if it's more confusing though (not that you'd be forced to use it)

I like the idea of this. It helps hide the h5 handle and makes for a decent whole-shot drop in replacement line. It does lend itself to strangeness like `run= Run(path); with run.open('r+') as shot: ...`, but some decent examples may help prevent that from happening.

 I'll go ahead and implement it as I think it is better.

File: docs/source/examples.rst
Comment (philipstarkey) [inline docs/source/examples.rst:70]: Needs to be `shot.`

File: lyse/__init__.py
Comment (philipstarkey) [inline lyse/__init__.py:399]: Should be `shot.`

File: lyse/__init__.py
Comment (philipstarkey) [inline lyse/__init__.py:384]: Like should end with a `:`?

**Review by philipstarkey — CHANGES_REQUESTED** [review]

File: docs/source/examples.rst
Comment (dihm) [reply docs/source/examples.rst]: This is what I get for typing and not checking...

**Comment by dihm** [comment]
Dumb typos fixed. Thanks for catching those @philipstarkey!

**Review by philipstarkey — APPROVED** [review]

**Pull request merged**
